compiler/rustc_data_structures/src/intern.rs

```diff
@@ -0,0 +1,102 @@
+use std::cmp::Ordering;
+use std::hash::{Hash, Hasher};
+use std::ops::Deref;
+use std::ptr;
+
+mod private {
+    #[derive(Clone, Copy, Debug)]
+    pub struct PrivateZst;
+}
+
+/// A reference to a value that is interned, and is known to be unique.
+///
+/// Note that it is possible to have a `T` and a `Interned<T>` that are (or
+/// refer to) equal but different values. But if you have two different
+/// `Interned<T>`s, they both refer to the same value, at a single location in
+/// memory. This means that equality and hashing can be done on the value's
+/// address rather than the value's contents, which can improve performance.
+///
+/// The `PrivateZst` field means you can pattern match with `Interned(v, _)`
+/// but you can only construct a `Interned` with `new_unchecked`, and not
+/// directly.
+#[derive(Debug)]
+#[cfg_attr(not(bootstrap), rustc_pass_by_value)]
+pub struct Interned<'a, T>(pub &'a T, pub private::PrivateZst);
+
+impl<'a, T> Interned<'a, T> {
+    /// Create a new `Interned` value. The value referred to *must* be interned
+    /// and thus be unique, and it *must* remain unique in the future. This
+    /// function has `_unchecked` in the name but is not `unsafe`, because if
+    /// the uniqueness condition is violated condition it will cause incorrect
+    /// behaviour but will not affect memory safety.
+    #[inline]
+    pub const fn new_unchecked(t: &'a T) -> Self {
+        Interned(t, private::PrivateZst)
+    }
+}
+
+impl<'a, T> Clone for Interned<'a, T> {
+    fn clone(&self) -> Self {
+        *self
+    }
+}
+
+impl<'a, T> Copy for Interned<'a, T> {}
+
+impl<'a, T> Deref for Interned<'a, T> {
+    type Target = T;
+
+    #[inline]
+    fn deref(&self) -> &T {
+        self.0
+    }
+}
+
+impl<'a, T> PartialEq for Interned<'a, T> {
+    #[inline]
+    fn eq(&self, other: &Self) -> bool {
+        // Pointer equality implies equality, due to the uniqueness constraint.
+        ptr::eq(self.0, other.0)
+    }
+}
+
+impl<'a, T> Eq for Interned<'a, T> {}
+
+impl<'a, T: PartialOrd> PartialOrd for Interned<'a, T> {
+    fn partial_cmp(&self, other: &Interned<'a, T>) -> Option<Ordering> {
+        // Pointer equality implies equality, due to the uniqueness constraint,
+        // but the contents must be compared otherwise.
+        if ptr::eq(self.0, other.0) {
+            Some(Ordering::Equal)
+        } else {
```

Comment:
I think you should be careful here as this implementation could be unsound. `PartialEq` is not reflexive. For something like `Interned<'a, f32>` somehow we have `Interned(&NaN) == Interned(&NaN)`.

I suggest using strict bounds: `T: Ord` instead because I do not think we have any interned type that only implements `PartialEq` but not `Eq` and I do not think we should have any in the future.

Reply:
Hmm. Values that are interned in a hash table (the standard way of doing it) require `Eq + Hash` because hash tables require that of their keys. So it's not possible to intern a float. Perhaps I should add `where Eq + Hash` to `Interned` to make things clearer?

With that in mind, is using strict bounds still a good idea? I'm having trouble understanding the problem case you are suggesting.

Reply:
My thoughts started from "can we reduce these two duplicate impls into one" and then came to conclusion that you should just use strict bounds here. Then you would not need to write a `PartialOrd` impl but only `Ord`.

Reply:
> That sounds like a good idea, since those are prerequisites for a type being interned, anyway.

I tried it, but then I had to add `where T: Eq + Hash` to every impl in the file, which was ugly, and then the compiler complained like so:
```
error[E0658]: trait bounds other than `Sized` on const fn parameters are unstable
  --> compiler/rustc_data_structures/src/intern.rs:29:10
   |
29 | impl<'a, T> Interned<'a, T> where T: Eq + Hash {
   |          ^
...
36 |     pub const fn new_unchecked(t: &'a T) -> Self {
   |     -------------------------------------------- function declared as const here
   |
   = note: see issue #57563 <https://github.com/rust-lang/rust/issues/57563> for more information
   = help: add `#![feature(const_fn_trait_bound)]` to the crate attributes to enable
```
and so I think I'll not do that after all :)

Reply:
> My thoughts started from "can we reduce these two duplicate impls into one" and then came to conclusion that you should just use strict bounds here. Then you would not need to write a `PartialOrd` impl but only `Ord`.

Sorry, I still don't follow. If I remove `PartialOrd` the compiler complains because you can't impl `Ord` without `PartialOrd`. Also, the two impls are very similar but not identical, because the return types are different and one calls `partial_cmp` while the other calls `cmp`. Perhaps I've misunderstood something?

Reply:
By bounding `T: Ord` on the `PartialOrd` `impl`, you can reuse the `Ord` `impl` by implementing `partial_cmp` as `Some(self.cmp(other))`

Reply:
I see, thanks for the explanation. I have updated the final commit to do this.